vars/yamlSubstituteVariables.groovy

```diff
@@ -0,0 +1,314 @@
+            return
+        }
+
+        if (!variablesFile.exists()) {
+            echo "[YamlSubstituteVariables] Could not find variable substitution file at ${variablesFilePath}. Skipping variable substitution."
+            return
+        }
+
+        def manifestData = null;
+        try {
+            // may return a List<Object>  (in case more YAML segments are in the file)
+            // or a Map<String, Object> in case there is just one segment.
+            manifestData = readYaml file: manifestFilePath
+            echo "[YamlSubstituteVariables] Loaded manifest at ${manifestFilePath}!"
+        }
+        catch(Exception ex) {
+            debug("Exception: ${ex}")
+            error "[YamlSubstituteVariables] Could not load manifest file at ${manifestFilePath}."
+            error "[YamlSubstituteVariables] Exception was: ${ex}"
+        }
+
+        def variablesData = null
+        try {
+            // may return a List<Object>  (in case more YAML segments are in the file)
+            // or a Map<String, Object> in case there is just one segment.
+            variablesData = readYaml file: variablesFilePath
+            echo "[YamlSubstituteVariables] Loaded variables file at ${variablesFilePath}!"
+        }
+        catch(Exception ex) {
+            debug("Exception: ${ex}")
+            error "[YamlSubstituteVariables] Could not load manifest variables file at ${variablesFilePath}."
+            error "[YamlSubstituteVariables] Exception was: ${ex}"
+        }
+
+        // substitute all variables.
+        ExecutionContext context = new ExecutionContext()
+        def result = substitute(manifestData, variablesData, context)
+
+        if (context.noVariablesReplaced) {
+            echo "[YamlSubstituteVariables] No variables were found or could be replaced in ${manifestFilePath}. Skipping variable substitution."
+            return
+        }
+
+        // writeYaml won't overwrite the file. You need to delete it first.
+        deleteOutputFileIfPresent(outputFilePath, skipDeletion)
+
+        writeYaml file: outputFilePath, data: result
+
+        echo "[YamlSubstituteVariables] Replaced variables in ${manifestFilePath} with variables from ${variablesFilePath}."
+        echo "[YamlSubstituteVariables] Wrote output file (with variables replaced) at ${outputFilePath}."
+
+        debug("Loaded Manifest: ${manifestData}")
+        debug("Loaded Variables: ${variablesData}")
+        debug("Result: ${result}")
+    }
+}
+
+/**
+ * Removes the given output file, if it exists.
+ * The {@code writeYaml} command won't delete files, so you need to
+ * remove the file yourself, in case it already exists.
+ * @param outputFilePath the path to the output file to write to.
+ */
+private void deleteOutputFileIfPresent(String outputFilePath, Boolean skipDeletion) {
+    File originalFile = new File(outputFilePath)
+    if (originalFile.exists() && !skipDeletion) {
```

Comment:
What happens later in case the file exists and `skipDeletion` is `true`? In this case the file does not get deleted here, but we fail later since the file is already present. That is weird.

Reply:
The `skipDeletion` flag is a nuisance, and I am not happy with it.
It actually is only to be used during unit testing, and its only purpose is to make sure that the test resources are not actually deleted / overwritten by the unit test.
Otherwise the unit test would always modify the (re)sources. 
It's really kind of difficult to test.
So consider the `skipDeletion` flag as a flag to be used by people (developers) who "know what they are doing", i.e. an internal mechanism.
If there is a better proposal, we can change it, of course.

Reply:
It is clear that test coding affects the productive code under tests (e.g. how methods are cut etc. But having such kind of boolean flags in the productive code for controlling method behaviour in tests is dangerous. It is present in the productive code and any developer after you might use it ...


> If there is a better proposal

I would like to suggest to try to adjust the corresponding method calls with `registerAllowedMethod`.  There are a lot of examples how that is done in the tests of that lib.

> "know what they are doing"

It is very dangerous to write code with that habit. That provides more options for failures for colleagues joining the project after you left and also for contributors from the outside world. Code is better when there is a lower degree on relationsship knowledge (ok, there is a flag, but it must only be used under certain circumstances ...). We can assume colleagues "know what they are doing" (to some exend) wrt language features (e.g. how to deal with multithreading, coercing ...), but we should not have that kind of relationship knowledge as a prerequisite IMO.

Reply:
I totally agree with you. I will have a look at it again, and see if I can somehow do it differently.

Reply:
Has been resolved by introducing a new new `deleteFile` script which is (re-)used inside of the `cfManifestSubstituteVariables` script and mocked via a new `JenkinsDeleteFileRule` which I also introduced. This rule can be configured NOT to delete files, as is required by some tests.